MOOSE Training and New Workshop - Intro, Tutorial Header, and Step 1-5

crswong888 opened the pull request:
# Updated Description (7/30/20)

This PR reorganizes the "Getting Started" section of the website to include all examples and all future tutorials as part of this procedure intended for new MOOSE users. The idea is that the reader proceed in the following order: "Install MOOSE" -> "New Users" -> "Examples and Tutorials."

This PR moves the `examples/` directory under `getting_strarted/examples_and_tutorials/examples/`, and moves its index content to `getting_started/examples_and_tutorials/index.md`. The header page, "Examples and Tutorials," includes an introduction, an "Examples" section, a "Tutorials" section, and a "Workshop" section to present the respective materials available for all three of these training types. A scrollbox MooseDocs feature was created under the Layout Extension to help format this page.

The "New Users" section includes everything from the old "Getting Started" page, except the install and update MOOSE instructions. I've included a section explaining how to enable physics modules in the applications `Makefile`. I've also added a section, "Helpful Software," to recommend that new users consider installing Atom, PEACOCK, and/or Paraview.

I've begun to develop a novel tutorial: "Tutorial 1: Application Development", which resembles the classic `darcy_thermo_mech` tutorial from the workshop. A link to this tutorial has been added to the "Examples and Tutorials" page. So far I have completed steps 1-5 of this tutorial. I created the necessary scripting in MooseDocs python to include content from `idaholab/babbler` needed for the tutorial demonstrations. 

I also created a new help page: "What is the Finite Element Method?" For which, I reference in the FAQ, the "Finite Element Concepts" page, and the tutorial. The discussion included serves the purpose of describing the core philosophy of the FEM, which is necessary for MOOSE developers to understand.

I needed to move `large_media/tutorials/` to `large_media/tutorials/darcy_thermo_mech/` so that I may distinguish media files associated with different tutorials, e.g., `large_media/tutorials/tutorial01_app_development/step02_result.png` and `large_media/tutorials/darcy_thermo_mech/step02_result.png`. To be clear, these two files are distinct - I actually cropped `step02_result.png` for Tutorial 1 so that it lays out more neatly on a MooseDown page. This change is reflected in my [`large_media:workshop-15232`](https://github.com/idaholab/large_media/tree/workshop-15232) branch. I updated all calls to the Media Extension in the workshop docs to be `!media darcy_thermo_mech/file` instead of `!media file`.

## Updated Description (6/30/20)

This PR initializes a novel approach to MOOSE new user training. The classic `darcy_thermo_mech` tutorial that is, typically, presented for MOOSE Workshops, shall be converted into a MooseDown format that appears much like the rest of the website. It shall be filed under a new section of the website: "Training", which will appear as a new mega menu. The current MOOSE Examples shall also appear within this new section. All future tutorials shall be appended to this portion of MOOSE.

The source code for tutorials and examples shall be located at `moose/training/tutorials` and `moose/training/examples`, respectively. Similarly, any media required for MooseDown content shall be filed under `large_media/training/tutorials` or `large_media/training/examples`. Finally, all new website content shall be filed under `modules/doc/training/*`. 

After most steps of the New Workshop tutorial, the reader will be asked to reproduce the problem themselves and commit their work to their own repositories. A new idaholab repository, `babbler`, was created to serve as the answer key for these demonstrations. It shall be added to MOOSE as a set of submodules - one for each commit, so that its entire history exists simultaneously and can be referenced as part of the Workshop content.

## Description (5/15/20)

Wrote the introductory potion of, as well as Step 1 of the `darcy_thermo_mech` tutorial for, the new MOOSE workshop - which converts the usual live workshops, supplemented with slides, into a virtual, MOOSEDown format that appears much like the rest of the website.

All of this is a major WIP, but this first merge will allow us to buff out some basic formatting issues before we start cranking out more pages. The final organization of the directories and their docs should be settled here.

(refs #15232 and #13617)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/527592/) on ae24840 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15300/site)

This comment will be updated on new commits.

aeslaughter commented on modules/doc/content/application_development/workshop/preface.md:
```diff
@@ -0,0 +1,36 @@
+
+There are many approaches to the FEM. One of the most common is the [Direct Stiffness Method](https://en.wikipedia.org/wiki/Direct_stiffness_method), which is probably what you were first taught when you took your Introduction to Finite Element Analysis course. This method usually involves constructing a system of linear equations in the following form,
+
+\begin{equation}
+\label{stiffness}
+\bold{A} x = b
+\end{equation}
```
```suggestion
!equation id=stiffness
\bold{A} x = b
```

aeslaughter replied:
Please use this syntax for equations

aeslaughter reviewed (commented):
This is a great start. I am going to push some re-organization changes and then provide some general comments.

aeslaughter commented:
Please move the examples to training/examples, remove the links to examples and the workshop in the main Documentation tab, and then create a new "Examples" tab that will contain links to the pages you are working on.

aeslaughter commented on modules/doc/content/training/darcy_thermo_mech/step01.md:
```diff
@@ -0,0 +1,124 @@
+# Step 1: Creating and Executing a MOOSE Object
```
Can you rename this file to step01_moose_object.md. I just like the filename giving some information about what it contains.

aeslaughter commented:
Another thought I have is that we should probably make the tutorial text complete from an application development standpoint. Step 01 should ask the user to create a github account, run stork, compile, run_tests, and push a commit. Then step02, create a problems directory, add a folder for diffusion, build input file, run it. Step 03, create a test for the input file you just made (syntax check in problems) and a smaller version in tests directory?

crswong888 commented:
> Another thought I have is that we should probably make the tutorial text complete from an application development standpoint. Step 01 should ask the user to create a github account, run stork, compile, run_tests, and push a commit. Then step02, create a problems directory, add a folder for diffusion, build input file, run it. Step 03, create a test for the input file you just made (syntax check in problems) and a smaller version in tests direc

crswong888 commented:
@aeslaughter I made the changes you requested and some of the other ones we talked about. Please take a moment to read through my Step 1 page so we can discuss how we want to integrate more hands-on stuff. In the meantime, I will look into the issue of expanding the contents extension to create a TOC with an outline style.

aeslaughter commented on modules/doc/content/training/darcy_thermo_mech/step01_moose_object.md:
```diff
@@ -0,0 +1,124 @@
+# Step 1: Creating and Executing a MOOSE Object
```
The MooseObject stuff should come in step 2, before the first Kernel is created. This step should focus on Diffusion and the input file.

crswong888 replied:
The reason I included that first was because I though it related well to input file syntax. Perhaps I should only briefly mention them and then move this more in-depth discussion to step 2. Although, I wanted to talk about Galerkin FEM and weak forms in step 2, so that step would become cluttered. Let me know what you think of this.

aeslaughter replied:
Step 2 can just be MooseObject; we can inherit from Diffusion originally so can leave all the FEM stuff for step3, which can use KernelGrad as a base class. This is just an idea, what do you think?

crswong888 replied:
I like this idea, we should spread out concept discussions as much as possible. 

I'm thinking I subdivide FEM even further, and only talk about weighted residuals and generating a weak form in step 3. Then, maybe step 4 can be about polynomial fitting, test functions = shape functions in Galerkin FEM, jacobians, etc. Finally step 5 can be about integration, i.e., Gauss quad rules. If I have to create all new tutorial steps to make this work I will.

Anyways, first I'll work on step 2 with MOOSE objects, then we'll worry about FEM steps.

crswong888 commented:
@aeslaughter I've pretty much addressed your comments, and now I'm sort of just proceeding along with more workshop content. Things are starting to get messy, and I have unfinished things lurking everywhere, but its fine, I'm keeping track of this.

We need to consider if we want to shift the focus of this workshop to the user building their own application. I foresee an approach where we create an entirely new repo under idaholab account, and make each step a commit, or its own branch, but still require the reader make their own repo and reproduce ours. Let me know what you think of this. 

Either way, I'm going to need to create all new files and directories, because my steps no longer correspond with those in the `tutorials/darcy_thermo_mech` directory.

aeslaughter commented:
> We need to consider if we want to shift the focus of this workshop to the user building their own application.

This should be the focus, I would like your steps to finish with git commands, to commit and push the content.

> I foresee an approach where we create an entirely new repo under idaholab account, and make each step a commit, or its own branch, but still require the reader make their own repo and reproduce ours. Let me know what you think of this.

We tried this, it is far too difficult to test and maintain. If you need to make a change to an input then you need to go into each commit and make the change. At least with the folders you can search and replace, plus we also use symlinks when the files do not change.

> Either way, I'm going to need to create all new files and directories, because my steps no longer correspond with those in the `tutorials/darcy_thermo_mech` directory.

Just make a new folder and leave the old one, we can remove the old one when you are done. This will also allow us to start merging content and testing what you have. What name do you have in mind for the folder?

crswong888 commented:
@aeslaughter I see the issue with creating a separate repository. I'll just create my own stork app on the side so that I can be sure my tutorial instructions are 100% reproducible.

That's a good question about the name of the directory. Since we're sort of consolidating all training materials under this new "Training" alias, what if I created a directory like `training/tutorials/darcy_thermo_mech`, and we can put all future tutorials there as well. We could also move the `examples` folder to `training/examples` as well, and I can fix anything that breaks as a result. I don't see a reason to call it anything except `darcy_thermo_mech`, it just needs a root thats different from the `tutorials` one.

I guess I will need to change many things I've done so far to follow the procedure of building a custom application. I will make this my primary focus now.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit c872d09b9ff18776002fcd0513ee5beae8c3e1ae
+Subproject commit e03d77223b44c440d8107f299ed20008826155a4
```
**Caution!** This contains a submodule update

crswong888 replied:
This is fine

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit c872d09b9ff18776002fcd0513ee5beae8c3e1ae
+Subproject commit e03d77223b44c440d8107f299ed20008826155a4
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit c7ff589edbb84aca6b140c463e38c941b37f4e2f
+Subproject commit e03d77223b44c440d8107f299ed20008826155a4
```
**Caution!** This contains a submodule update

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit c7ff589edbb84aca6b140c463e38c941b37f4e2f
+Subproject commit d14e29284b90c90636de0e733d8fb67523b59dc5
```
**Caution!** This contains a submodule update

crswong888 committed:
Revised "getting started" section of website to consist of 3 parts: installation, new users, and examples & tutorials

(refs #15232)

crswong888 committed:
Update links to "getting started" throughout website pages

(refs #15232)

crswong888 committed:
Moved "examples" under "examples and tutorials"

(refs #15232)

crswong888 commented:
This pull request was doing some weird things when comparing the diff. Please excuse all the opening and closing and base switching

crswong888 committed:
Added !content pagination to examples

(refs #15232)

crswong888 committed:
Started "Tutorial 1: Application Development" - created links to landing page

(refs #15232)

crswong888 committed:
Created "Problem Statement" page, added bib reference, and helpful acronyms

(refs #15232)

crswong888 committed:
Created a "What is the FEM?" page as a helpful discussion to be referenced wherever needed

(refs #15232)

crswong888 committed:
Created pages for tutorial 1: steps 1-5, as well as a placeholder page for step 6

(refs #15232)

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit a8826a77056184b2bc2ebe0ec78db0f5ed2455bc
+Subproject commit 86bc31d21d61a21b9ac297f71ce4f3611866d232
```
**Caution!** This contains a submodule update

crswong888 commented:
@aeslaughter I've totally reorganized the commit history here to make it easier for you to review and for anyone else to follow now, or in the future. At each commit, I confirmed that the website builds with no errors. Thus, the HEAD at each commit is a working version of MooseDocs. I stored a backup of the old commit history on my branch [`crswong888:workshop-15232-old`](https://github.com/crswong888/moose/tree/workshop-15232-old). The heads of these two branches are almost identical, but the "old" one has a history that is a lot messier.

I updated the description for this PR, hopefully, for the last time. I will not work develop this PR any further. My goal now is to get this merged. I will branch off and continue working on tutorial content there. Besides addressing your imminent change requests, the only thing that is left to do is to checkout `large_media` at `master`. Currently, all of my media content exists on a branch of `large_media`, which I will merge into `master` once this PR has your approval.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit a8826a77056184b2bc2ebe0ec78db0f5ed2455bc
+Subproject commit a2cb1b27ae876f805d6e0f5e9a1e9234052e5d54
```
**Caution!** This contains a submodule update

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit a8826a77056184b2bc2ebe0ec78db0f5ed2455bc
+Subproject commit 3d1c2cc8c69c88f029e37e3eb8fbceeaaa74542d
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

crswong888 committed:
Update links to `darcy_thermo_mech` media. Also corrected the boundary value problem given on Step 1 of workshop.

(refs #15232)

crswong888 commented:
@aeslaughter looks like I broke MASTODON docs because it also has a MooseDown page named `problem_statement.md`. All I'd have to do is add a bit more verbose to the filename, i.e., `!include isoil/problem_statement.md`. 

Although, I'm not sure what the protocol is for fixing a MOOSE application when you break it in the framework. Should I just open a PR in MASTODON and reference this PR?

crswong888 commented:
Alright I made a PR in mastodon to address the issue with the `problem_statement.md` filename: idaholab/mastodon#321.

crswong888 commented:
New branch to continue developing more tutorial content: [`crswong888:tutorial01-15232`](https://github.com/crswong888/moose/tree/tutorial01-15232)

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit a8826a77056184b2bc2ebe0ec78db0f5ed2455bc
+Subproject commit 50e26f0c41f4bcdb62b00b064dd07f8f128e0cd0
```
**Caution!** This contains a submodule update

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit a8826a77056184b2bc2ebe0ec78db0f5ed2455bc
+Subproject commit d34856046a6fd5ab7eba829b18b75f4dedbaffaa
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

aeslaughter commented on python/MooseDocs/extensions/layout.py:
```diff
@@ -89,6 +92,19 @@ def createToken(self, parent, info, page):
 
         return col
 
+class ScrollCommand(command.CommandComponent):
```
I am not sold on this command, can you pull this stuff out into another branch/PR? Don't do that until I push up my other changes.

crswong888 replied:
@aeslaughter Yes. Just let me know when you are done.

aeslaughter committed:
Re-work and simplify some text

(refs #15300)

aeslaughter commented:
@crswong888 I made quite a few changes, but don't let it be discouraging, what you have done is good and the whole concept is taking shape. This is a huge change and I expect some back and forth on the text as we finish it up.

I have four more files to go through but wanted to get you some of my changes and talk to you about it. **Thanks for all this work**, this is a big and important step to making our docs much better.

aeslaughter committed:
Edit new_users.md

(refs #15300)

aeslaughter committed:
Step 4

(refs #15300)

aeslaughter committed:
Step 05

(refs #15300)

crswong888 committed:
created Babbler application directories which represent a repo HEAD following each tutorial step

(refs #15232)

crswong888 commented:
After a strong effort to make it work, it became clear that adding `idaholab/babbler` as a submodule is not practical from a maintenance standpoint. I dropped the two commits which handled the Babbler stuff and replaced them with a commit that adds the applications as simple directories in `tutorials/tutorial01_app_development/`.

I'm going to continue to maintain `idaholab/babbler` via my personal repository [`crswong888/babbler_devel`](https://github.com/crswong888/babbler_devel), because I've made this is a very convenient method to handle all of the folders like a Git repo. I can modify the folders, first, in `crswong888/babbler_devel`, then push them as commits to `idaholab/babbler` to verify that everything looks good, and finally use [`babbler_devel/scripts/copy_to_moose.sh`](https://github.com/crswong888/babbler_devel/blob/master/scripts/copy_to_moose.sh) to copy them to `moose/tutorials/tutorial01_app_development/` and add tracking.

When this project nears completion, we must discuss what to do with `idaholab/babbler`. Perhaps we create a shell script to update this repository in `idaholab/moose` that Civet runs whenever changes are made to files matching `moose/tutorials/tutorial01_app_development/*`. This way, we can reference a real, exemplary, repository in the tutorial for users to compare their own work to. Thus, preserving the original intent of creating Babbler.

crswong888 commented:
Moved the scrollbox commit to #15760

crswong888 committed:
Edit new_users, examples_and_tutorials, tutorial01_app_development, and step01_moose_app

(refs #15232)

crswong888 committed:
Made it possible to hide outline indices or skip counting them altogether

crswong888 committed:
This collapsible-sections stuff looks really nice! It's just impossible to use unless we add it to the config token

crswong888 committed:
changed Acronym extension in such a way that it becomes possible to build a list of acronyms used across a set of pages

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 5300b0c16175d4aed7ae3568987fc0f6f28a0fee
+Subproject commit 0d1dc8e692c22ff282953a7f107fd94ac31782d6
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

crswong888 committed:
New !content outline settings in tutorial01

crswong888 committed:
Added a "Preface" page to include instructions for navigating the tutorial, an acronyms list, and any other general information that we may decide to include

(refs #15232)

crswong888 committed:
new `!acronym list location` setting in tutorial01

crswong888 commented:
@aeslaughter I couldn't resist: I've proposed some significant back-end MooseDocs changes here. Any commit labeled "WIP" is work done in Python.

I am continuing to make changes to actual tutorial content.

crswong888 committed:
Changed to AD formulations wherever appropriate.

Note: you can use scripts available from 'git@github.com:crswong888/babbler_devel.git' to manage these Babbler app directories much like a single git repository.

(refs #15232)

crswong888 committed:
Edit Step 2: Input File

(refs #15232)

crswong888 committed:
Edit Step 3: Moose Object

(refs #15232)

crswong888 committed:
Edit Step 4: Weak Form

(refs #15232)

crswong888 committed:
Edit Step 5: Kernel Object

(refs #15232)

crswong888 commented:
@aeslaughter Please see the changes I have pushed when you get the chance.

crswong888 committed:
Dropped the "Statement of Physics" headers and made some minor edits in Tutorial 1

(refs #15232)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/584366/) on f91e7ec wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15300/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15300/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 321d58ad0aac6464b3cdd7aac5a5a8aa559b8a8d`

crswong888 committed:
Clang formats in Babbler applications

(refs #15232)

crswong888 commented:
@aeslaughter okay, we could initiate an update for the build/execution system and move the commit I have here to a separate branch, specifically for this. And then, open an issue for a MooseDocs executioner overhaul. This ticket would then basically just sit on the back burner until the Tutorial is complete, at least.

crswong888 commented:
@aeslaughter

crswong888 committed:
Added table of kernel methods that evaluate residuals and edited some body text

(refs #15232)

aeslaughter reviewed (changes requested):
This is amazing! I will be adding a commit with a few minor changes to the text.

The only thing I want to change is the build.py changes. If you want to improve build.py, then I would like the file timestamp based approach implemented that we talked about. I can help you with the design of it.

aeslaughter commented:
You will also need to rebase and fix the conflicts. After you do that, could you also add a few sentences to the Nov. newsletter about the content and what else is coming. You won't need to worry about doing that later then.

aeslaughter committed:
Minor change to new tutorial content

(refs #15300)

aeslaughter commented on modules/doc/content/getting_started/examples_and_tutorials/tutorial01_app_development/step01_moose_app.md:
```diff
@@ -84,7 +84,7 @@ The stork script initializes a new application with a single git commit whose me
 cd ~/projects/babbler
 git remote add origin https://github.com/YourGitHubUserName/babbler
-git push -u origin master
+git push -u origin main
```
GitHub now defaults to "main"

crswong888 committed:
Resolving upstream conflicts

crswong888 committed:
Merged new media into large_media/master and checked out the submodule at master

crswong888 committed:
Minor edits to tutorial content

(refs #15232)

crswong888 committed:
Added a newsletter announcement for the release of this tutorial

(refs #15232)

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 5300b0c16175d4aed7ae3568987fc0f6f28a0fee
+Subproject commit fae3bd6561bb3dd59b397c462321809aea724bce
```
**Caution!** This contains a submodule update

crswong888 commented:
@aeslaughter I'm not sure if I'm the best person to compose an official newsletter announcement, but I've at least laid out the basic message.

The most recent commit should not actually be merged. I attempted to fix some broken links with the large_media check out, hoping it would work, which it didn't. But, I didn't want to just throw this away. Ultimately, this commit can be cherry-picked to save some time.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 0ef1a6c3ac0dff740f02edb3d9ed75c7181ecba2
+Subproject commit fae3bd6561bb3dd59b397c462321809aea724bce
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

aeslaughter committed:
Update SQA traceability for changes to status checks

(refs #15232)

aeslaughter commented:
BISON fix is on its way, this created an incompatible docs situation
